package/ota_tool/src/main.rs

```diff
@@ -0,0 +1,160 @@
+                     .map_err(|e| format!("Error encoding reboot output: {}", e.to_string()))?;
+    println!("{}", cmd_output);
+    return Ok(());
+}
+
+fn is_uuid(uuid: String) -> Result<(), String> {
+    match Uuid::parse_str(&uuid) {
+        Err(_) => Err(String::from("Device UUID not correctly formatted")),
+        Ok(_) => Ok(())
+    }
+}
+
+fn ota_service_request(uuid: Uuid, current_version: &str, url: &str) -> Result<OTAServiceResponse, String> {
+    let mut response = reqwest::Client::new()
+        .get(url)
+        .header(DeviceUid(uuid))
+        .header(CurrentVersion(String::from(current_version)))
+        .send().map_err(|e| format!("Error making request to OTA service: {}", e))?;
+    if !response.status().is_success() {
+        return Err(format!("OTA service request error, server responded {}", response.status()));
+    }
+    return response.json().map_err(|e| format!("Error parsing JSON response: {}", e));
+}
+
+fn main() {
+    // Parse command line arguments
+    let matches = App::new("ota_tool")
+        .about("Over-the-Air Update (OTA) Tool. Checks for new firmware and installs it.")
+        .author("Swift Navigation")
+        .arg(Arg::with_name("UUID")
+                 .help("Device UUID")
+                 .required(true)
+                 .index(1)
+                 .validator(is_uuid))
+        .arg(Arg::with_name("VERSION")
+                 .help("Current Firmware Version")
+                 .required(true)
+                 .index(2))
+        .arg(Arg::with_name("url")
+                 .short("u")
+                 .long("url")
+                 .value_name("URL")
+                 .help("Service endpoint URL")
+                 .takes_value(true))
+        .get_matches();
+
+    // OTA service endpoint URL
+    let url = matches.value_of("url")
+                     .unwrap_or("https://upgrader.skylark.swiftnav.com/images");
+
+    // OTA request parameters
+    let current_version = matches.value_of("VERSION")
+                                 .expect("Error retrieving command line argument VERSION");
+    let uuid = Uuid::parse_str(matches.value_of("UUID").unwrap())
+                    .expect("Error parsing UUID");
+
+    println!("endpoint url: {}", url);
+    println!("device uuid: {}", uuid);
+    println!("current version: {}", current_version);
+
+    // Make a request to the OTA service to get the details of what firmware we
+    // should be running
+    let decoded = ota_service_request(uuid, current_version, url).expect("OTA service request error");
+
+    println!("target version: {}", decoded.version);
+
+    if decoded.version == current_version {
+      println!("Already up-to-date");
+      return;
+    }
+
+    println!("url: {}", decoded.url);
+    println!("expected sha256: {}", decoded.sha256);
+
+    let mut fw_req = BufReader::new(reqwest::get(&decoded.url).expect("Error downloading firmware"));
+    let fw_path = Path::new("/tmp/PiksiMulti.bin");
```

Comment:
We should write the upgrade `.bin` to `/data` since it has a block of memory reserved for upgrade data.